app/components/webview/DAppWebView.tsx

```
@@ -192,6 +197,73 @@ export const DAppWebView = memo(forwardRef((props: DAppWebViewProps, ref: Forwar
+                            dispatchWalletResponse(ref as RefObject<WebView>, { result: false });
+                            return;
+                        }
+                        (async () => {
+                            const result = await WalletService.checkIfCardIsAlreadyAdded(primaryAccountNumberSuffix);
+                            dispatchWalletResponse(ref as RefObject<WebView>, { result });
+                        })();
+                        break;
+                    case 'canAddCard':
+                        const cardId = parsed.data.args.cardId;
+                        if (!cardId) {
+                            warn('Invalid cardId');
+                            dispatchWalletResponse(ref as RefObject<WebView>, { result: false });
+                            return;
+                        }
+
+                        (async () => {
+                            const result = await WalletService.canAddCard(cardId);
+                            dispatchWalletResponse(ref as RefObject<WebView>, { result });
+                        })();
+                        break;
+                    case 'addCardToWallet':
+                        const request = addCardRequestSchema.safeParse(parsed.data.args);
+
+                        if (!request.success) {
+                            warn('Invalid addCardToWallet request');
+                            dispatchWalletResponse(ref as RefObject<WebView>, { result: false });
+                            return;
+                        }
+
+                        const token = getHoldersToken(getCurrentAddress().address.toString({ testOnly: isTestnet }));
+
+                        if (!token) {
+                            warn('User token not found');
+                            dispatchWalletResponse(ref as RefObject<WebView>, { result: false });
+                            return;
+                        }
+
+                        (async () => {
```

Promise failure is not handled here